proposals/NNNN-package-manager-product-definitions.md

```diff
@@ -7,135 +7,123 @@
-            ])
-        ],
-    )
-    ```
+```
+Target(name: "Foo", dependencies: ["Bar", "Baz"])
+```
     
-    In this example, the package name is the short name of the package as defined by the package itself.  A possibility would be to allow the package parameter to be optional when it is unambiguous.
+is shorthand for:
     
-    A significant drawback of this approach is the loss of an ability to use strings as unambiguous shorthand for `.target(name: "...")` values.  To remedy this, strings could still be allowed, as long as there was a clear way to disambiguate the dependency in cases in which two or more entities have the same name.
+```
+Target(name: "Foo", dependencies: [.target(name: "Bar"), .target(name: "Baz")])
+```
+
+This will be extended to support product dependencies in addition to target dependencies:
     
-    If we look at practical use cases, it seems rare that two completely unrelated entities (e.g. a target and a product) would have the same name.  For example, if a target depends on `libAlamofire`, it doesn't actually matter whether it's the target or the product that is the dependency.  So in most cases, a string would unambiguously capture the intent of the package author.  If the shorthand form followed a search order, such as:
+```
+let package = Package(
+    name: "MyClientLib",
+    dependencies: [
+        .package(url: "https://github.com/Alamofire/Alamofire", majorVersion: 3),
+    ],
+    targets: [
+        Target(name: "MyUtils"),
+        Target(name: "MyClientLib", dependencies: [
+            .target(name: "MyUtils"),
+            .product(name: "Alamofire", package: "Alamofire")
+        ])
+    ]
+)
+```
     
-    - first look for a target with the specified name in the same package
-    - second, look for a product with the specified name in any package dependency
+The package name is the canonical name of the package, as defined in the manifest of the package that defines the product.  The product name is the name specified in the product definition of that same manifest.
 
-    then it should be possible for the shorthand notation to be sufficient in the vast majority of cases, with the possibility of using the more verbose form in the cases in which the shorthand form really is ambiguous.
+In order to continue supporting the convenience of being able to use plain strings as shorthand, and in light of the fact that most of the time the names of packages and products are unique enough to avoid confusion, we will extend the short-hand notation so that a string can refer to either a target or a product.
 
-2.  Alternatively, a new optional array parameter could be added to the instantiation of a target:
+The Package Manager will first try to resolve the name to a target in the same package; if there isn't one, it will instead to resolve it to a product in one of the packages specified in the `dependencies` parameter of the `Package()` initializer.
```

Comment:
Actually:  is this needed, conceptually, if we also let you specify the package name to disambiguate?  I know that there are currently technical limitations that would cause problems, but do you think there are also any inherent conceptual problems with that?